src/hotspot/share/opto/superword.cpp

```diff
@@ -2950,177 +3002,110 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) {
     if (in_bb(n)) {
+      block_count++;
+
       if (n->is_LoadStore() || n->is_MergeMem() ||
           (n->is_Proj() && !n->as_Proj()->is_CFG())) {
         // Bailout if the loop has LoadStore, MergeMem or data Proj
         // nodes. Superword optimization does not work with them.
+#ifndef PRODUCT
+      if (is_trace_superword_any()) {
+        tty->print_cr("SuperWord::construct_bb: fails because of unhandled node:");
+        n->dump();
+      }
+#endif
         return false;
       }
-      bb_ct++;
+
+#ifdef ASSERT
       if (!n->is_CFG()) {
         bool found = false;
         for (uint j = 0; j < n->req(); j++) {
           Node* def = n->in(j);
-          if (def && in_bb(def)) {
+          if (def != nullptr && in_bb(def)) {
             found = true;
             break;
           }
         }
-        if (!found) {
-          assert(n != entry, "can't be entry");
-          _data_entry.push(n);
-        }
-      }
-    }
-  }
-
-  // Find memory slices (head and tail)
-  for (DUIterator_Fast imax, i = lp()->fast_outs(imax); i < imax; i++) {
-    Node *n = lp()->fast_out(i);
-    if (in_bb(n) && n->is_memory_phi()) {
-      Node* n_tail  = n->in(LoopNode::LoopBackControl);
-      if (n_tail != n->in(LoopNode::EntryControl)) {
-        if (!n_tail->is_Mem()) {
-          assert(n_tail->is_Mem(), "unexpected node for memory slice: %s", n_tail->Name());
-          return false; // Bailout
-        }
-        _mem_slice_head.push(n);
-        _mem_slice_tail.push(n_tail);
+        assert(found, "every non-cfg node must have an input that is also inside the loop");
       }
+#endif
     }
   }
 
-  // Create an RPO list of nodes in block
-
+  // Create a reverse-post-order list of nodes in block
   ResourceMark rm;
   GrowableArray<Node*> stack;
   VectorSet visited;
   VectorSet post_visited;
 
-  // Push all non-control nodes with no inputs from within block, then control entry
-  for (int j = 0; j < _data_entry.length(); j++) {
-    Node* n = _data_entry.at(j);
-    visited.set(bb_idx(n));
-    stack.push(n);
-  }
-  visited.set(bb_idx(entry));
-  stack.push(entry);
+  visited.set(bb_idx(bb()));
+  stack.push(bb());
 
   // Do a depth first walk over out edges
-  int rpo_idx = bb_ct - 1;
-  int size;
-  int reduction_uses = 0;
-  while ((size = stack.length()) > 0) {
+  int rpo_idx = block_count - 1;
+  while (!stack.is_empty()) {
     Node* n = stack.top(); // Leave node on stack
     if (!visited.test_set(bb_idx(n))) {
       // forward arc in graph
     } else if (!post_visited.test(bb_idx(n))) {
       // cross or back arc
+      int old_length = stack.length();
       for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
-        Node *use = n->fast_out(i);
+        Node* use = n->fast_out(i);
         if (in_bb(use) && !visited.test(bb_idx(use)) &&
             // Don't go around backedge
-            (!use->is_Phi() || n == entry)) {
-          if (is_marked_reduction(use)) {
```

Comment:
Note: I would like to remove this reduction code from the basic block code. It separates the different "components".

Instead, we can just check `is_marked_reduction_loop` outside, which checks if we have any marked reduction.
The only difference is that we don't do the implemented check. But most platforms implement the reductions, and we check that again later in `SuperWord::implemented` anyway.